---
layout: post
title: "PRS score (2)"
subtitle: "PRS calculation and presentation"
date: 2022-05-26 10:24:48
header-style: text
catalog: true
author: "Yuan"
tags: [PLINK, PRSice-2, LDPred-2, lassosum]
---
{% include linksref.html %}

>君子博学而日参省乎己，则知明而行无过矣

This is part 2 of PRS score calculation blog. The [first part](https://raymondshang.github.io/2022/05/24/prs-score/) focuses on QC in base and target GWAS

## Required files
Before calculating PRS scores, the following files are required:

| File Name    |                                                    Description |
| :----------- | -------------------------------------------------------------: |
| Height.QC.gz |                                The post-QCed summary Statistic |
| EUR.QC.bed   |        The genotype file after performing some basic filtering |
| EUR.QC.bim   |    This file contains the SNPs that passed the basic filtering |
| EUR.QC.fam   | This file contains the samples that passed the basic filtering |
| EUR.height   |                This file contains the phenotype of the samples |
| EUR.cov      |               This file contains the covariates of the samples |

## Update Effect Size
When the effect size relates to disease risk and is thus given as an odds ratio (OR), rather than BETA (for continuous traits), then the PRS is computed as a product of ORs. To simplify this calculation, we take the natural logarithm of the OR so that the PRS can be computed using summation instead (which can be back-transformed afterwards).
```r
library(data.table)
dat <- fread("Height.QC.gz")
fwrite(dat[,BETA:=log(OR)], "Height.QC.Transformed", sep="\t")
q() # exit R
```
> :warning: Due to rounding of values, using awk to log transform OR can lead to less accurate results. 

## Clumping
Linkage disequilibrium, which corresponds to the correlation between the genotypes of genetic variants across the genome, makes identifying the contribution from causal independent genetic variants extremely challenging. One way of approximately capturing the right level of causal signal is to perform clumping, which removes SNPs in ways that only weakly correlated SNPs are retained but preferentially retaining the SNPs most associated with the phenotype under study. Clumping can be performed using the following command in [plink](https://www.cog-genomics.org/plink/1.9/postproc#clump):
```bash
plink \
    --bfile EUR.QC \
    --clump-p1 1 \
    --clump-r2 0.1 \
    --clump-kb 250 \
    --clump Height.QC.Transformed \
    --clump-snp-field SNP \
    --clump-field P \
    --out EUR
```

parameters used above:

| Parameter       |         Value         | Description                                                                                                           |
| :-------------- | :-------------------: | :-------------------------------------------------------------------------------------------------------------------- |
| clump-p1        |           1           | P-value threshold for a SNP to be included as an index SNP. 1 is selected such that all SNPs are include for clumping |
| clump-r2        |          0.1          | SNPs having $r^2$ higher than 0.1 with the index SNPs will be removed                                                 |
| clump-kb        |          250          | SNPs within 250k of the index SNP are considered for clumping                                                         |
| clump           | Height.QC.Transformed | Base data (summary statistic) file containing the P-value information                                                 |
| clump-snp-field |          SNP          | Specifies that the column SNP contains the SNP IDs                                                                    |
| clump-field     |           P           | Specifies that the column P contains the P-value information                                                          |
>:note:The $r^2$ values computed by --clump are based on maximum likelihood haplotype frequency estimates

The EUR.clumped files contain the index SNPs, which could be extract:
```bash
awk 'NR!=1{print $3}' EUR.clumped >EUR.valid.snp
```
> :note: If your target data are small (e.g. N < 500) then you can use the 1000 Genomes Project samples for the LD calculation. Make sure to use the population that most closely reflects represents the base sample.

## Generate PRS using plink
plink provides a convenient function --score and --q-score-range for calculating polygenic scores.
Three files are required here:
1. The base data file: Height.QC.Transformed
2. A file containing SNP IDs and their corresponding P-values
```bash
awk '{print $3,$8}' Height.QC.Transformed > SNP.pvalue
```
3. A file containing the different P-value thresholds for inclusion of SNPs in the PRS. 

```bash
echo "0.001 0 0.001" > range_list 
echo "0.05 0 0.05" >> range_list
echo "0.1 0 0.1" >> range_list
echo "0.2 0 0.2" >> range_list
echo "0.3 0 0.3" >> range_list
echo "0.4 0 0.4" >> range_list
echo "0.5 0 0.5" >> range_list
```
The format of range_list is:
```bash
Name_of_Threshold	Lower_bound	Upper_Bound
```

### calculating PRS using plink

```bash
plink \
    --bfile EUR.QC \
    --score Height.QC.Transformed 3 4 12 header \
    --q-score-range range_list SNP.pvalue \
    --extract EUR.valid.snp \
    --out EUR
```

|Parameter|Value|Description|
|score|Height.QC.Transformed 3 4 12 header|We read from the Height.QC.Transformed file, assuming that the 3st column is the SNP ID; 4th column is the effective allele information; the 12th column is the effect size estimate; and that the file contains a header|
|q-score-range|range_list SNP.pvalue|We want to calculate PRS based on the thresholds defined in range_list, where the threshold values (P-values) were stored in SNP.pvalue|

The above command and range_list will generate 7 files:
1. EUR.0.5.profile
2. EUR.0.4.profile
3. EUR.0.3.profile
4. EUR.0.2.profile
5. EUR.0.1.profile
6. EUR.0.05.profile
7. EUR.0.001.profile

{{callout_info}} 
PRS is calculated in plink as below:<br/><br/>

$$
    PRS_j =\frac{ \sum_i^NS_i*G_{ij}}{P*M_j}
$$

<br/><br/>Where the effect size of SNP $i$ is $S_i$; the number of effect alleles observed in sample $j$ is $G_{ij}$; the ploidy of the sample is $P$ (is genrally 2 for human); the total number of SNPs included in the PRS is N;and the number of non-missing SNPs observed in sample $j$ is $M_j$. If the sample has a missing genotype for SNP $i$, then the population minor allele frequency multiplied by the ploidy ($MAF_i * P$) is used instead of $G_{ij}$.
{{end}}

